commands/src/main/java/org/wildfly/extras/creaper/commands/deployments/Deploy.java

```diff
@@ -0,0 +1,211 @@
+package org.wildfly.extras.creaper.commands.deployments;
+
+import org.jboss.as.controller.client.OperationBuilder;
+import org.jboss.dmr.ModelNode;
+import org.wildfly.extras.creaper.core.online.Constants;
+import org.wildfly.extras.creaper.core.online.OnlineCommand;
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.DEPLOY;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.ENABLED;
+
+/**
+ * Command which takes care about deploying specified deployment under specified deployment name.
+ * <p/>
+ * In case of domain it is deployed by default to all server groups
+ */
+public final class Deploy implements OnlineCommand {
+
+    private final String deploymentName;
+    private final InputStream deploymentInputStream;
+    private final boolean autoCloseInputStream;
+    private final List<String> serverGroups;
+
+    private Deploy(InputStream deploymentInputStream, Data deployCmdOptions) {
+        this.deploymentInputStream = deploymentInputStream;
+        this.deploymentName = deployCmdOptions.deploymentName;
+        this.autoCloseInputStream = deployCmdOptions.autoCloseInputStream;
+        this.serverGroups = deployCmdOptions.serverGroups;
+    }
+
+    @Override
+    public void apply(OnlineCommandContext ctx) throws Exception {
+        Operations ops = new Operations(ctx.client);
+
+        // deploy and add to server groups
+        ModelNode content = new ModelNode();
+        content.get(Constants.INPUT_STREAM_INDEX).set(0);
+
+        ModelNode deployRequest;
+
+        if (ctx.options.isDomain) {
+            List<String> groups = serverGroups;
+            if (serverGroups == null)
+                groups = ops.readChildrenNames(Address.root(), Constants.SERVER_GROUP)
+                        .stringListValue(Collections.<String>emptyList());
+
+            List<ModelNode> groupDeploymentAddresses = new ArrayList<ModelNode>(groups.size());
+            for (String serverGroup : groups) {
+                ModelNode mn = new ModelNode();
+                mn.add(Constants.SERVER_GROUP, serverGroup);
+                mn.add(Constants.DEPLOYMENT, deploymentName);
+                groupDeploymentAddresses.add(mn);
+            }
+            deployRequest = createDeploymentOperation(content,
+                    groupDeploymentAddresses.toArray(new ModelNode[groupDeploymentAddresses.size()]));
```

Comment:
Aha, the `Operation` thing. So far, I'm only aware of a single thing that needs to attach streams to operations: deployments. Let's stick with `ModelNode`s here for now.